text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,236 @@
+
+This way of representing the new messages has been choose over introducing an enum wrapper to minimize breaking changes of XCM message decoding in tools like Subscan for example. 
+
+Example: 
+```
+[ XCM message1, XCM message2, ..., EMPTY message, UMPSignal::CoreSelector ]
+```
+
+#### `UMPSignal` messages
+
+```
+/// A strictly increasing sequence number, tipically this would be the parachain block number.
+pub struct CoreSelector(pub BlockNumber);
+
+/// An offset in the relay chain claim queue.
+pub struct ClaimQueueOffset(pub u8);
+
+/// Default claim queue offset
+pub const DEFAULT_CLAIM_QUEUE_OFFSET: ClaimQueueOffset = ClaimQueueOffset(1);
+
+pub enum UMPSignal {
+	/// A message sent by a parachain to select the core the candidate is commited to.
+	/// Relay chain validators, in particular backers, use the `CoreSelector` and `ClaimQueueOffset`
+	/// to compute the index of the core the candidate has commited to.
+	///
+	SelectCore(CoreSelector, ClaimQueueOffset),
+}
+```
+
+As we dont want to have a claim queue snapshot in the parachain runtime, we need to set `ClaimQueueOffset` 
+statically to some sane value. Parachains should prefer to have a static value that makes sense for their usecase which can be changed by governance at some future point. Changing the value dynamically can be a friction point. It will work out fine to decrease the value to build more into the present. But if the value is increased to build more into the future, a relay chain block will be skipped.
+
+Considering `para_assigned_cores` is a sorted vec of core indices assigned to a parachain at the
+specified claim queue offset, validators will determine the committed core index like this:
+
+```
+let assigned_core_index = core_selector % para_assigned_cores.len();
+let committed_core_index = para_assigned_cores[assigned_core_index];
+```
+
+
+### Polkadot Primitive changes
+
+#### New [CandidateDescriptor](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/primitives/src/v7/mod.rs#L482)
+
+- reclaim 32 bytes from `collator: CollatorId` and 64 bytes from `signature: CollatorSignature` and rename to `reserved1`  and `reserved2` fields.
+- take 2 bytes from `reserved1` for a new `core_index: u16` field.  
+- take 4 bytes from `reserved2` for a new `session_index: u32` field.
+- the `reserved1` and `reserved2` fields are zeroed
+
+Thew new primitive will look like this:
+```
+pub struct CandidateDescriptorV2<H = Hash> {
+	/// The ID of the para this is a candidate for.
+	para_id: Id,
+	/// The hash of the relay-chain block this is executed in the context of.
+	relay_parent: H,
+	/// The core index where the candidate is backed.
+	core_index: CoreIndex,
+	/// The session index in which the candidate is backed.
+	session_index: SessionIndex,
```

Comment:
uh ok. They are actually next to each other. Then I am assuming above is a typo. Swapping the two might be a good idea: core index is 2 bytes, session index 4. This way we should end up with better alignment.

Reply:
The encoded layout is different than the Rust memory layout which Rust might not obey the ordering in the type declaration.